node/manager: prevent MeshNodeSync from pruning cluster (local) nodes

0xch4z opened the pull request:
Fixes an issue in NodeManager where restored cluster nodes can be pruned before the initial node listing completes.

Currently, with clustermesh, when all remote clusters have finished initial synchronization, `MeshNodeSync` is called, which considers all mesh and non-mesh nodes for pruning. If this happens before we finish listing initial cluster (local) nodes, we could end up considering the restored cluster nodes as stale for pruning. This also causes some miscounting for metrics, where for example we decrement `nodes_all_num` (in `NodeDeleted` for the synthetic delete event) for nodes we never accounted for in the first place.

This patch modifies MeshNodeSync to only consider pruning mesh nodes.

```release-note
Fixes an issue in NodeManager where restored cluster nodes can be pruned before the initial node listing completes.
```

0xch4z committed:
node/manager: prevent MeshNodeSync from pruning cluster (local) nodes

Fixes an issue in NodeManager where restored cluster nodes can be pruned
before the initial node listing completes.

Currently, with clustermesh, when all remote clusters have finished
initial synchronization, `MeshNodeSync` is called, which considers all
mesh and non-mesh nodes for pruning. If this happens before we finish
listing initial cluster (local) nodes, we could end up considering the
restored cluster nodes as stale for pruning. This also causes some
miscounting for metrics, where for example we decrement `nodes_all_num`
(in `NodeDeleted` for the synthetic delete event) for nodes we never
accounted for in the first place.

This patch modifies MeshNodeSync to only consider pruning mesh nodes.

Signed-off-by: Charlie Kenney <charles.kenney@isovalent.com>

bimmlerd commented on pkg/node/manager/manager.go:
```diff
@@ -1170,11 +1170,14 @@ func (m *manager) NodeSync() {
 }
 
-func (m *manager) pruneNodes(includeMeshed bool) {
+func (m *manager) pruneNodes(filterMeshed bool) {
```
From what I can tell, we now have two variants of `pruneNodes` without overlap - one which looks at cluster-local ones and one which looks at meshed nodes. Can we make this explicit with two methods (maybe `pruneLocalNodes` and `pruneMeshNodes`)? It might be less DRY but I feel like the clarity of intent comes through more than a boolean argument.

0xch4z replied:
I went with `pruneClusterNodes` since `local` is used a lot in the comments to refer to the node the agent is running on, but happy to change it to local if that seems preferable!

bimmlerd reviewed (changes requested):
looks like a nice find! One code suggestion from my side, but cc @giorio94 who may have more insight

0xch4z committed:
node/manager: only decrement num_nodes_all for nodes we've added

DeleteNodes is called with synthetic delete events for nodes that have
been deleted between NodeManager restarts. Now, we check that the given
node is not a stale node being pruned (that we haven't accounted for)
before we decrement num_nodes_all.

Signed-off-by: Charlie Kenney <charles.kenney@isovalent.com>

giorio94 commented on pkg/node/manager/manager_test.go:
```diff
@@ -1349,27 +1349,41 @@ func TestNodeIpset(t *testing.T) {
 func TestNodesStartupPruning(t *testing.T) {
 	logger := hivetest.Logger(t)
-	n1 := nodeTypes.Node{Name: "node1", Cluster: "c1", IPAddresses: []nodeTypes.Address{
+	c1Node1 := nodeTypes.Node{Name: "c1-node1", Cluster: "c1", IPAddresses: []nodeTypes.Address{
```
Nit. It may be worthy having the node names overlap across the two clusters, just as an extra check that everything works correctly in that case as well, and the cluster name is taken into account.

giorio94 commented on pkg/node/manager/manager_test.go:
```diff
@@ -1416,40 +1430,50 @@ func TestNodesStartupPruning(t *testing.T) {
+	// current view).
+	// Restored cluster nodes should not be pruned yet.
+	mngr.NodeUpdated(c2Node1)
+	mngr.MeshNodeSync()
```
Nit. I could make sense testing the opposite case as well, in which cluster nodes are synced first, followed by clustermesh synchronization. Just to prevent future regressions in that scenario.

giorio94 reviewed (approved):
Thanks @0xch4z for finding and fixing this! The changes look good to me, just a couple of nits inline about the test.

joamaki commented on pkg/node/manager/manager.go:
```diff
@@ -1185,34 +1189,85 @@ func (m *manager) pruneNodes(includeMeshed bool) {
-		if n.Cluster == m.conf.ClusterName || includeMeshed {
-			m.NodeDeleted(*n)
-			delete(m.restoredNodes, id)
+	for _, n := range toDelete {
```
Could the fact that we're calling this after releasing the mutex be causing https://github.com/cilium/cilium/issues/41543? E.g. we call `pruneClusterNodes` before local node has been added and thus plan to synthesize the delete for it and then when we release the mutex the local node gets added and then we get here and immediately delete it?
